lib/utils/ignore.ts

```diff
@@ -174,3 +190,136 @@ export class FileIgnorer {
+the ".dockerignore" file. There are however subtle differences in how the patterns
+are interpreted between the .gitignore and .dockerignore files, and you are
+encouraged to review the patterns to suit your project. Dockerignore reference:
+https://docs.docker.com/engine/reference/builder/#dockerignore-file
+--------------------------------------------------------------------------------`;
+		console.error(msg);
+		return;
+	}
+
+	// Check the ".dockerignore" file for required entries. Each pattern line is compared
+	// disregarding the first character if it is a '!' or '#', because the user is allowed
+	// to modify the patterns by respectively inverting them or commenting them out.
+
+	let dockerIgnoreStr: string;
+	try {
+		dockerIgnoreStr = await fs.readFile(dockerIgnorePath, 'utf8');
+	} catch (err) {
+		return exitWithExpectedError(
+			`Error reading file ${dockerIgnorePath}: ${err.message}`,
+		);
+	}
+	const dockerIgnoreLines: string[] = dockerIgnoreStr
+		.split('\n') // also OK if the separator is '\r\n' (Windows CR-LF)
+		.map(v => v.trim());
+	const dockerIgnoreLinesMinusPrefix: string[] = dockerIgnoreLines.map(v =>
+		v.startsWith('!') || v.startsWith('#') ? v.slice(1) : v,
+	);
+	const missingEntries = defaultEntries.filter(v => {
+		v = v.startsWith('!') ? v.slice(1) : v;
+		return !_.some(dockerIgnoreLinesMinusPrefix, p => p === v);
+	});
+
+	if (missingEntries.length) {
+		const entriesAre = missingEntries.length === 1 ? 'entry is' : 'entries are';
+		const msg = `\
+--------------------------------------------------------------------------------
+Warning: the following ${entriesAre} missing in the ".dockerignore" file:
+
+${missingEntries.join(os.EOL)}
+
+${dockerIgnoreMsg}
+--------------------------------------------------------------------------------`;
+		console.error(msg);
+	}
+}
+
+/**
+ * Create and return a Transform stream suitable for a '.pipe(stream)' operation,
+ * for use with the 'klaw' package. The stream will use a '.dockerignore' file
+ * (if any) to filter out files and folders.
+ * @param projectDir The project source folder ('-s' command-line option)
+ */
+export async function getKlawStreamFilterForDockerIgnore(
+	projectDir: string,
+): Promise<Transform> {
+	const dockerIgnorePath = path.join(projectDir, '.dockerignore');
+	let dockerIgnoreStr = '';
+	try {
+		dockerIgnoreStr = await fs.readFile(dockerIgnorePath, 'utf8');
+	} catch (err) {
+		if (err.code !== 'ENOENT') {
+			return exitWithExpectedError(
+				`Error reading file "${dockerIgnorePath}": ${err.message}`,
+			);
+		}
+	}
+	const through2 = await import('through2');
+	const dockerIgnore = require('@zeit/dockerignore')();
+	if (dockerIgnoreStr) {
+		dockerIgnore.add(dockerIgnoreStr);
+	}
+	// Don't ignore any metadata files
+	// The pattern below matches `.balena/qemu` and `myservice/.balena/qemu`
+	// but not `some.dir.for.balena/qemu`.
+	dockerIgnore.add(['!**/.balena', '!**/.resin']);
```

Comment:
A test is failing on Windows and I have traced it to a bug in the `@zeit/dockerignore` library:
https://github.com/zeit/dockerignore/issues/15

The library bug seems to be actually worse than I reported on that issue: it seems that any pattern including a slash (forward slash or backslash) fails to match on Windows. This also affects the existing dockerignore implementation in the CLI, but could become a more acute problem if we remove support for gitignore.